docs/udf.rst

```diff
@@ -0,0 +1,367 @@
+.. _ksql-udfs:
+
+KSQL User Defined Function Reference
+====================================
+
+KSQL supports creating User Defined Functions (UDFs) via custom jars that are
+uploaded to the ext/ directory of the KSQL installation. Scalar and aggregate functions
+are supported.
+At start up time KSQL scans the jars in the directory looking for any classes that annotated
+with ``@UdfDescription`` (scalar functions) or ``@UdafDescription`` (aggregate functions).
+Classes annotated with ``@UdfDescription`` are scanned for any public methods that are annotated
+with ``@Udf``. Classes annotated with ``@UdafDescription`` are scanned for any public static methods
+that are annotated with ``@UdafFactory``. Each UDF that is found is parsed and, if successful, loaded into KSQL.
+
+Each UDF instance has its own child-first ``ClassLoader`` that is isolated from other UDFs. If you
+need to use any third-party libraries with your UDFs then they should also be part of your jar, i.e.,
+you should create an "uber-jar". The classes in your uber-jar will be loaded in preference to any
+classes on the KSQL classpath excluding anything vital to the running of KSQL, i.e., classes that are
+part of ``org.apache.kafka`` and ``io.confluent``. Further, the ``ClassLoader`` can restrict access
+to other classes via a blacklist. The blacklist file is ``resource-blacklist.txt``. You can add
+any classes or packages that you want blacklisted from UDF use, for example you may not
+want a UDF to be able to fork processes. Further details on how to blacklist are available below.
+
+=============
+Creating UDFs
+=============
+
+Scalar UDF
+----------
+
+To create a Scalar UDF you need to create a class that is annotated with ``@UdfDescription``.
+Each method in the class that represents a UDF must be public and annotated with ``@Udf``. The class
+you create represents a collection of scalar UDFs all with the same name but may have different
+arguments and return types.
+
+
+Example Scalar UDF class
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The class below creates a UDF named ``multiply``. As can be seen this UDF can be invoked in three
```

Comment:
Can we highlight that:

- the UDF name that one must use in KSQL is determined by the `name` parameter in `@UdfDescription` and `@UdafDescription`, respectively.
- the UDF name, when used in KSQL statements, is case-insensitive.  so `multiply` and `MULTIPLY` both work, for example.

We can also fold this into the table that describes `@UdfDescription` in the next section.